File: bgpc/src/cpp_simulation_framework/as_graph_analyzer.cpp
#include <stack>
#include "announcement.hpp"
#include "as.hpp"
#include "cpp_simulation_engine.hpp"
#include "as_graph_analyzer.hpp"


ASGraphAnalyzer::ASGraphAnalyzer(std::shared_ptr<CPPSimulationEngine> engine,
                             const std::vector<unsigned short int>& ordered_prefix_block_ids,
                             const std::unordered_set<int>& victim_asns,
                             const std::unordered_set<int>& attacker_asns,
                             bool data_plane_tracking,
                             bool control_plane_tracking)
    : engine(engine),
      ordered_prefix_block_ids(ordered_prefix_block_ids),
      victim_asns(victim_asns),
      attacker_asns(attacker_asns),
      data_plane_tracking(data_plane_tracking),
      control_plane_tracking(control_plane_tracking){

    // Reserve sizes for speed
    control_plane_outcomes.reserve(ordered_prefix_block_ids.size());
    data_plane_outcomes.reserve(ordered_prefix_block_ids.size());
    /*
    most_specific_ann_dict.reserve(engine->as_graph->as_dict.size());

    for (auto& as_obj_pair : engine->as_graph->as_dict) {
        auto& as_obj = as_obj_pair.second;
        most_specific_ann_dict[as_obj->asn] = get_most_specific_ann(as_obj, ordered_prefix_block_ids);
    }
    */
}

std::unordered_map<int, std::unordered_map<int, int>> ASGraphAnalyzer::analyze() {
    for (auto& as_obj_pair : engine->as_graph->as_dict) {
        auto& as_obj = as_obj_pair.second; // Assuming as_graph is a map from IDs to AS shared_ptrs

        // Data Plane Analysis
        if(data_plane_tracking){
            int data_outcome = get_as_outcome_data_plane(as_obj);
            data_plane_outcomes[as_obj->asn] = data_outcome;
        }

        // Control Plane Analysis
        if (control_plane_tracking){
            int ctrl_outcome = get_as_outcome_ctrl_plane(as_obj);
            control_plane_outcomes[as_obj->asn] = ctrl_outcome;
        }
    }

    // Using integer values from the Plane enum as keys
    outcomes[static_cast<int>(Plane::DATA)] = data_plane_outcomes;
    outcomes[static_cast<int>(Plane::CTRL)] = control_plane_outcomes;
    return outcomes;
}


std::shared_ptr<Announcement> ASGraphAnalyzer::get_most_specific_ann(std::shared_ptr<AS> as_obj, const std::vector<unsigned short int>& ordered_prefix_block_ids) {

    std::shared_ptr<Announcement> most_specific_ann = nullptr;
    for (const auto& prefix_block_id : ordered_prefix_block_ids) {
        most_specific_ann = as_obj->policy->localRIB.get_ann(prefix_block_id);
        if (most_specific_ann) {
            return most_specific_ann;
        }
    }
    return nullptr;  // Return empty optional if no announcement is found
}



int ASGraphAnalyzer::get_as_outcome_data_plane(const std::shared_ptr<AS>& as_obj) {
    // Check if the outcome is already determined for the AS
    auto it = data_plane_outcomes.find(as_obj->asn);
    if (it != data_plane_outcomes.end()) {
        return it->second;
    }

    // Get the most specific announcement for the AS
    // std::shared_ptr<Announcement> most_specific_ann = most_specific_ann_dict[as_obj->asn];
    // NOTE: doing this here and avoiding a dict makes this 1.5x faster at least
    // This is essential for the python runtimes
    std::shared_ptr<Announcement> most_specific_ann = nullptr;
    for (const auto& prefix_block_id : ordered_prefix_block_ids) {
        most_specific_ann = as_obj->policy->localRIB.get_ann(prefix_block_id);
        if (most_specific_ann) {
            break;
        }
    }


    // Determine the outcome based on the most specific announcement
    int outcome = determine_as_outcome_data_plane(as_obj, most_specific_ann);

    // Handle undetermined cases, potentially involving recursion
    if (outcome == static_cast<int>(Outcomes::UNDETERMINED) && most_specific_ann) {
        auto next_as = engine->as_graph->as_dict[most_specific_ann->as_path[1]]; // Example of getting the next AS in the path
        outcome = get_as_outcome_data_plane(next_as); // Recursively determine the outcome
    }

    assert(outcome != static_cast<int>(Outcomes::UNDETERMINED)); // Ensure that the outcome is no longer undetermined

    // Store and return the determined outcome
    data_plane_outcomes[as_obj->asn] = outcome;
    return outcome;
}

int ASGraphAnalyzer::determine_as_outcome_data_plane(const std::shared_ptr<AS>& as_obj, const std::shared_ptr<Announcement>& most_specific_ann) {
    // Check if the AS is an attacker
    if (attacker_asns.find(as_obj->asn) != attacker_asns.end()) {
        return static_cast<int>(Outcomes::ATTACKER_SUCCESS);
    }

    // Check if the AS is a victim
    if (victim_asns.find(as_obj->asn) != victim_asns.end()) {
        return static_cast<int>(Outcomes::VICTIM_SUCCESS);
    }

    // Check if there is no announcement or other specific conditions are met
    if (!most_specific_ann ||
        most_specific_ann->as_path.size() == 1 ||
        most_specific_ann->recv_relationship == Relationships::ORIGIN ||
        most_specific_ann->traceback_end) {
        return static_cast<int>(Outcomes::DISCONNECTED);
    }

    // If none of the above conditions are met, the outcome is undetermined
    return static_cast<int>(Outcomes::UNDETERMINED);
}

int ASGraphAnalyzer::get_as_outcome_ctrl_plane(std::shared_ptr<AS> as_obj) {
    // Check if the outcome is already computed for the AS
    auto it = control_plane_outcomes.find(as_obj->asn);
    if (it != control_plane_outcomes.end()) {
        return it->second;
    }

    // Get the most specific announcement for the AS
    // std::shared_ptr<Announcement> most_specific_ann = most_specific_ann_dict[as_obj->asn];
    std::shared_ptr<Announcement> most_specific_ann = nullptr;
    for (const auto& prefix_block_id : ordered_prefix_block_ids) {
        most_specific_ann = as_obj->policy->localRIB.get_ann(prefix_block_id);
        if (most_specific_ann) {
            break;
        }
    }


    // Determine the outcome based on the most specific announcement
    int outcome = determine_as_outcome_ctrl_plane(as_obj, most_specific_ann);

    // Store and return the determined outcome
    control_plane_outcomes[as_obj->asn] = outcome;
    return outcome;
}

int ASGraphAnalyzer::determine_as_outcome_ctrl_plane(std::shared_ptr<AS> as_obj, std::shared_ptr<Announcement> ann) {
    // If there is no announcement, the AS is considered disconnected in the control plane
    if (!ann) {
        return static_cast<int>(Outcomes::DISCONNECTED);
    }

    // Check if the origin of the announcement is an attacker
    if (attacker_asns.find(ann->origin()) != attacker_asns.end()) {
        return static_cast<int>(Outcomes::ATTACKER_SUCCESS);
    }

    // Check if the origin of the announcement is a victim
    if (victim_asns.find(ann->origin()) != victim_asns.end()) {
        return static_cast<int>(Outcomes::VICTIM_SUCCESS);
    }

    // If none of the above conditions are met, the AS is considered disconnected
    // This can be adjusted based on more specific rules for your scenario
    return static_cast<int>(Outcomes::DISCONNECTED);
}

int ASGraphAnalyzer::get_other_as_outcome_hook(std::shared_ptr<AS> as_obj) {
    // Used to satisfy type checker, hook method for other metrics
    return static_cast<int>(Outcomes::ATTACKER_SUCCESS);
}

/*
// Using a stack and avoiding recursion to make it faster
int ASGraphAnalyzer::get_as_outcome_data_plane(std::shared_ptr<AS> initial_as_obj) {
    std::stack<std::shared_ptr<AS>> as_stack;
    as_stack.push(initial_as_obj);

    while (!as_stack.empty()) {
        auto as_obj = as_stack.top();
        as_stack.pop();

        // Check if the outcome is already determined for the AS
        auto it = data_plane_outcomes.find(as_obj->asn);
        if (it != data_plane_outcomes.end()) {
            continue;
        }

        // Get the most specific announcement for the AS
        // std::shared_ptr<Announcement> most_specific_ann = most_specific_ann_dict[as_obj->asn];
        // NOTE: doing this here and avoiding a dict makes this 1.5x faster at least
        // This is essential for the python runtimes
        std::shared_ptr<Announcement> most_specific_ann = std::nullopt;
        for (const auto& prefix_block_id : ordered_prefix_block_ids) {
            std::shared_ptr<Announcement> most_specific_ann = as_obj->policy->localRIB.get_ann(prefix_block_id);
            if (most_specific_ann) {
                break;
            }
        }


        // Determine the outcome based on the most specific announcement
        int outcome = determine_as_outcome_data_plane(as_obj, most_specific_ann);

        if (outcome == static_cast<int>(Outcomes::UNDETERMINED) && most_specific_ann) {
            auto next_as = engine->as_graph->as_dict[(*most_specific_ann)->as_path[1]];
            as_stack.push(next_as); // Push the next AS to the stack instead of a recursive call
        } else {
            assert(outcome != static_cast<int>(Outcomes::UNDETERMINED)); // Ensure that the outcome is no longer undetermined
            data_plane_outcomes[as_obj->asn] = outcome; // Store the determined outcome
        }
    }

    // Since we are using a stack, the initial AS's outcome should be in the data_plane_outcomes map
    return data_plane_outcomes[initial_as_obj->asn];
}
*/

File: bgpy/simulation_framework/as_graph_analyzers/as_graph_analyzer.py
from typing import TYPE_CHECKING, Optional

from bgpy.as_graphs import AS
from bgpy.shared.enums import Outcomes, Plane, Relationships
from bgpy.simulation_engine import Announcement as Ann
from bgpy.simulation_engine import BaseSimulationEngine

from .base_as_graph_analyzer import BaseASGraphAnalyzer

if TYPE_CHECKING:
    from bgpy.simulation_framework.scenarios import Scenario


class ASGraphAnalyzer(BaseASGraphAnalyzer):
    """Takes in a BaseSimulationEngine and outputs metrics"""

    def __init__(
        self,
        engine: BaseSimulationEngine,
        scenario: "Scenario",
        data_plane_tracking: bool = True,
        control_plane_tracking: bool = False,
        # Most to least specific
        ordered_prefixes: tuple[str, ...] = (),
    ) -> None:
        self.engine: BaseSimulationEngine = engine
        self.scenario: Scenario = scenario
        if not ordered_prefixes:
            ordered_prefixes = tuple(self.scenario.ordered_prefix_subprefix_dict.keys())
        self._most_specific_ann_dict: dict[AS, Ann | None] = (
            self._get_most_specific_ann_dict(engine, ordered_prefixes)
        )
        self._data_plane_outcomes: dict[int, int] = dict()
        self._control_plane_outcomes: dict[int, int] = dict()
        self.outcomes: dict[int, dict[int, int]] = {
            Plane.DATA.value: self._data_plane_outcomes,
            Plane.CTRL.value: self._control_plane_outcomes,
        }
        self.data_plane_tracking: bool = data_plane_tracking
        self.control_plane_tracking: bool = control_plane_tracking

    def _get_most_specific_ann_dict(
        self, engine: BaseSimulationEngine, ordered_prefixes: tuple[str, ...]
    ) -> dict[AS, Optional["Ann"]]:
        """Gets the most specific ann in a list of ordered prefixes

        ordered prefixes start with the most specific, and move to least specific
        """

        return {
            x: self._get_most_specific_ann(x, ordered_prefixes) for x in engine.as_graph
        }

    def _get_most_specific_ann(
        self, as_obj: AS, ordered_prefixes: tuple[str, ...]
    ) -> Optional["Ann"]:
        """Returns the most specific announcement that exists in a rib

        as_obj is the as
        ordered prefixes are prefixes ordered from most specific to least
        """

        for prefix in ordered_prefixes:
            most_specific_ann = as_obj.policy.local_rib.get(prefix)
            if most_specific_ann:
                assert isinstance(most_specific_ann, Ann), "for mypy"
                return most_specific_ann
        return None

    def analyze(self) -> dict[int, dict[int, int]]:
        """Takes in engine and outputs traceback for ctrl + data plane data"""

        for as_obj in self.engine.as_graph:
            if self.data_plane_tracking:
                # Gets AS outcome and stores it in the outcomes dict
                self._get_as_outcome_data_plane(as_obj)
            if self.control_plane_tracking:
                self._get_as_outcome_ctrl_plane(as_obj)
            self._get_other_as_outcome_hook(as_obj)
        return self.outcomes

    ####################
    # Data plane funcs #
    ####################

    def _get_as_outcome_data_plane(self, as_obj: AS) -> int:
        """Recursively returns the as outcome"""

        if as_obj.asn in self._data_plane_outcomes:
            return self._data_plane_outcomes[as_obj.asn]
        else:
            most_specific_ann = self._most_specific_ann_dict[as_obj]
            outcome_int = self._determine_as_outcome_data_plane(
                as_obj, most_specific_ann
            )
            # We haven't traced back all the way on the AS path
            if outcome_int == Outcomes.UNDETERMINED.value:
                assert most_specific_ann, "If not disconnected, ann must exist"
                # next as in the AS path to traceback to
                # Ignore type because only way for this to be here
                # Is if the most specific "Ann" was NOT None.
                next_as = self.engine.as_graph.as_dict[
                    # NOTE: Starting in v4, this is the next hop,
                    # not the next ASN in the AS PATH
                    # This is more in line with real BGP and allows for more
                    # advanced types of hijacks such as origin spoofing hijacks
                    most_specific_ann.next_hop_asn
                ]
                outcome_int = self._get_as_outcome_data_plane(next_as)
            assert outcome_int != Outcomes.UNDETERMINED.value, "Shouldn't be possible"

            self._data_plane_outcomes[as_obj.asn] = outcome_int
            return outcome_int

    def _determine_as_outcome_data_plane(
        self, as_obj: AS, most_specific_ann: Optional["Ann"]
    ) -> int:
        """Determines the outcome at an AS

        ann is most_specific_ann is the most specific prefix announcement
        that exists at that AS
        """

        if as_obj.asn in self.scenario.attacker_asns:
            return Outcomes.ATTACKER_SUCCESS.value
        elif as_obj.asn in self.scenario.victim_asns:
            return Outcomes.VICTIM_SUCCESS.value
        # End of traceback
        elif (
            most_specific_ann is None
            or len(most_specific_ann.as_path) == 1
            or most_specific_ann.recv_relationship.value == Relationships.ORIGIN.value
            # Adding this condition in V4 for proper next_hop behavior
            or most_specific_ann.next_hop_asn == as_obj.asn
        ):
            return Outcomes.DISCONNECTED.value
        else:
            return Outcomes.UNDETERMINED.value

    #######################
    # Control Plane Funcs #
    #######################

    def _get_as_outcome_ctrl_plane(self, as_obj: AS) -> int:
        """Stores and returns the AS outcome from the control plane"""

        most_specific_ann = self._most_specific_ann_dict[as_obj]
        outcome_int = self._determine_as_outcome_ctrl_plane(as_obj, most_specific_ann)
        assert outcome_int != Outcomes.UNDETERMINED.value, "Shouldn't be possible"
        self._control_plane_outcomes[as_obj.asn] = outcome_int
        return outcome_int

    def _determine_as_outcome_ctrl_plane(self, as_obj: AS, ann: Optional["Ann"]) -> int:
        """Determines the outcome at an AS on the control plane

        ann is most_specific_ann is the most specific prefix announcement
        that exists at that AS
        """

        if not ann:
            return Outcomes.DISCONNECTED.value
        elif ann.origin in self.scenario.attacker_asns:
            return Outcomes.ATTACKER_SUCCESS.value
        elif ann.origin in self.scenario.victim_asns:
            return Outcomes.VICTIM_SUCCESS.value
        else:
            return Outcomes.DISCONNECTED.value

    ################################
    # Hook funcs for other metrics #
    ################################

    def _get_other_as_outcome_hook(self, as_obj: AS) -> int:
        # Noop, this is just to satisfy mypy
        return Outcomes.ATTACKER_SUCCESS.value
